src/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs

```diff
@@ -0,0 +1,475 @@
+        }
+
+        /// <summary>
+        /// Actually preforms the transfer.
+        /// </summary>
+        /// <remarks>Does NO checks.</remarks>
+        /// <param name="to">Address of a new owner.</param>
+        /// <param name="tokenId">The NFT that is being transferred.</param>
+        private void TransferInternal(Address to, ulong tokenId)
+        {
+            Address from = GetIdToOwner(tokenId);
+            ClearApproval(tokenId);
+
+            RemoveNFToken(from, tokenId);
+            AddNFToken(to, tokenId);
+
+            LogTransfer(from, to, tokenId);
+        }
+
+        /// <summary>
+        /// Removes a NFT from owner.
+        /// </summary>
+        /// <remarks>Use and override this function with caution. Wrong usage can have serious consequences.</remarks>
+        /// <param name="from">Address from wich we want to remove the NFT.</param>
+        /// <param name="tokenId">Which NFT we want to remove.</param>
+        private void RemoveNFToken(Address from, ulong tokenId)
+        {
+            Assert(GetIdToOwner(tokenId) == from);
+            SetOwnerToNFTokenCount(from, checked(GetOwnerToNFTokenCount(from) - 1));
+            this.PersistentState.Clear(GetIdToOwnerKey(tokenId));
+        }
+
+        /// <summary>
+        /// Assignes a new NFT to owner.
+        /// </summary>
+        /// <remarks>Use and override this function with caution. Wrong usage can have serious consequences.</remarks>
+        /// <param name="to">Address to which we want to add the NFT.</param>
+        /// <param name="tokenId">Which NFT we want to add.</param>
+        private void AddNFToken(Address to, ulong tokenId)
+        {
+            Assert(GetIdToOwner(tokenId) == Address.Zero);
+
+            SetIdToOwner(tokenId, to);
+            ulong currentTokenAmount = GetOwnerToNFTokenCount(to);
+            SetOwnerToNFTokenCount(to, checked(currentTokenAmount + 1));
+        }
+
+        /// <summary>
+        /// Actually perform the safeTransferFrom.
+        /// </summary>
+        /// <param name="from">The current owner of the NFT.</param>
+        /// <param name="to">The new owner.</param>
+        /// <param name="tokenId">The NFT to transfer.</param>
+        /// <param name="data">Additional data with no specified format, sent in call to 'to' if it is a contract.</param>
+        private void SafeTransferFromInternal(Address from, Address to, ulong tokenId, byte[] data)
+        {
+            CanTransfer(tokenId);
+            ValidNFToken(tokenId);
+
+            Address tokenOwner = GetIdToOwner(tokenId);
+            Assert(tokenOwner == from);
+            Assert(to != Address.Zero);
+
+            TransferInternal(to, tokenId);
+
+            if (this.PersistentState.IsContract(to))
+            {
+                ITransferResult result = this.Call(to, 0, "OnNonFungibleTokenReceived", new object[] { this.Message.Sender, from, tokenId, data }, 0);
```

Comment:
Why does this return an object which I can't cast in any way too a bool?
I can also not case to Nullable<bool> to resolve the compiler error I get thus making it pretty hard to work with.
Thirdly System.Convert is not supported so I also have no way of converting this with a system class.